Configurable range for auto scale

dancek opened the pull request:
Scales below 1.0 are typically unuseful, so make 1.0 a minimum auto scale by default.

Add configuration options for minimum and maximum auto scale, in case someone actually wants scales under 1.0 or wishes to limit the maximum scale.

Fixes #135.

dancek commented:
I decided to make it configurable while I'm at it, and also add an optional maximum. (If I had access to multiple slightly over 200dpi displays I might well want to limit the maximum to 2.0.)

There's no command-line interface for adjusting the range as that feels like a rare operation.

alex-courtis commented:
> I decided to make it configurable while I'm at it, and also add an optional maximum. (If I had access to multiple slightly over 200dpi displays I might well want to limit the maximum to 2.0.)

Many thanks. We'll likely need max at some stage...

> There's no command-line interface for adjusting the range as that feels like a rare operation.

That's fine, it would indeed be rare.

alex-courtis commented on src/head.c:
```diff
@@ -154,8 +154,22 @@ wl_fixed_t head_auto_scale(struct Head *head) {
 	long dpi_quantized = (long)(dpi / 12 + 0.5) * 12;
 
+	// convert min and max to quantized dpi
+	long dpi_min = 12 * (long)(min * 8 + 0.9999f);
```
Why the 0.9999f ? Is it to do with the double parameters? Could they be floats?

dancek replied:
This was a quick-n-dirty implementation of `ceil`, ie. the value needs to be rounded up. I'll write something smarter instead.

dancek replied:
Changed to ceil() from math.h.

alex-courtis reviewed (changes requested):
Many thanks, love your work.

Testing:

cli / log
- [ ] shows in -g
- [x] shows in -y -g

min
- [ ] default - when min or max not present scale is always set to 1
- [x] above
- [x] below
- [x] bad
- [x] 0

max
- [ ] default - when min or max not present scale is always set to 1
- [x] above
- [x] below
- [x] 0

other
- [x] specific scale
- [x] scaling off
- [x] auto scaling off
- [x] change cfg

odd
- [ ] min > max - not quite predictable

alex-courtis commented on doc/CONFIGURATION.md:
```diff
@@ -104,6 +104,22 @@ This may be disabled and scale 1 will be used, unless a `SCALE` has been specifi
+
+### AUTO_SCALE_MAX
+
+Set maximum value for auto scaling. Values below 1.0 are ignored.
```
Can we just accept values below 1.0? There might be a use case...

dancek replied:
Yeah, I changed that.

dancek commented:
Thanks for a really good review! I'll work on these when I have time (hopefully in a couple of days) and get back to you!

dancek commented:
> cli / log
> 
> * [ ]  shows in -g

Added.

> min
> 
> * [ ]  default - when min not present scale is always set to 1

Actually this was when max wasn't set, due to accidentally using the default min value as the default max value too. Fixed.

> odd
> 
> * [ ]  min > max - not quite predictable

Fixed by disabling it in this case.

---

Give it another look with these fixes. I'll be happy to fix any further issues.

alex-courtis reviewed (approved):
Fantastic, all working nicely:

cli / log
- [x] shows in -g
- [x] shows in -y -g

min
- [x] default - when min not present scale is always set to 1 - fixed
- [x] above
- [x] below
- [x] bad
- [x] 0

max
- [x] default
- [x] above
- [x] below
- [x] bad
- [x] 0
- [x] 0.1 - goes to calculated

min > max
- [x] uses min

other
- [x] specific scale
- [x] scaling off
- [x] auto scaling off
- [x] change cfg

alex-courtis commented:
Many thanks for your contribution!

Can you wait for a release? I'd like to get #121 in as well...

dancek commented:
I'm in no hurry, take your time!
